test/dd/test_package.cpp

```diff
@@ -285,21 +285,86 @@ TEST(DDPackageTest, IdentityTrace) {
   auto mul = dd->multiply(tr, tr);
-  EXPECT_EQ(dd::RealNumber::val(mul.w.r), 4.0);
+  EXPECT_EQ(dd::RealNumber::val(mul.w.r), 1.);
 }
 
-TEST(DDPackageTest, PartialNonIdentityTrace) {
+TEST(DDPackageTest, PartialSWapMatTrace) {
   auto dd = std::make_unique<dd::Package<>>(2);
+  auto dd2 = std::make_unique<dd::Package<>>(2);
   auto swapGate = dd->makeTwoQubitGateDD(dd::SWAP_MAT, 0, 1);
+  auto swapGate2 = dd2->makeTwoQubitGateDD(dd::SWAP_MAT, 0, 1);
   auto ptr = dd->partialTrace(swapGate, {true, false});
-  EXPECT_EQ(ptr.w * ptr.w, 1.);
+  auto fullTrace = dd->trace(ptr, 1);
+  auto fullTraceOriginal = dd2->trace(swapGate2, 2);
+  EXPECT_EQ(dd::RealNumber::val(ptr.w.r), 0.5);
+  // Check that successively tracing out subsystems is the same as computing the
+  // full trace from the beginning
+  EXPECT_EQ(fullTrace.r, fullTraceOriginal.r);
+}
+
+TEST(DDPackageTest, PartialTraceKeepInnerQubits) {
+  // Check that the partial trace computation is correct when tracing out the
+  // outer qubits only. This test shows that we should avoid storing
+  // non-eliminated nodes in the compute table, as this would prevent their
+  // proper elimination in subsequent trace calls.
+
+  const std::size_t numQubits = 8;
+  auto dd = std::make_unique<dd::Package<>>(numQubits);
+  auto dd2 = std::make_unique<dd::Package<>>(numQubits);
+  const auto swapGate = dd->makeTwoQubitGateDD(dd::SWAP_MAT, 0, 1);
+  auto swapKron = swapGate;
+  for (std::size_t i = 0; i < 3; ++i) {
+    swapKron = dd->kronecker(swapKron, swapGate, 2);
+  }
+  const auto swapGate2 = dd2->makeTwoQubitGateDD(dd::SWAP_MAT, 0, 1);
+  auto swapKron2 = swapGate2;
+  for (std::size_t i = 0; i < 3; ++i) {
+    swapKron2 = dd2->kronecker(swapKron2, swapGate2, 2);
+  }
+  auto fullTraceOriginal = dd2->trace(swapKron2, numQubits);
+  auto ptr = dd->partialTrace(
+      swapKron, {true, true, false, false, false, false, true, true});
+  auto fullTrace = dd->trace(ptr, 4);
+  EXPECT_EQ(dd::RealNumber::val(ptr.w.r), 0.25);
+  EXPECT_EQ(fullTrace.r, 0.0625);
+  // Check that successively tracing out subsystems is the same as computing the
+  // full trace from the beginning
+  EXPECT_EQ(fullTrace.r, fullTraceOriginal.r);
+}
+
+TEST(DDPackageTest, TraceComplexity) {
+  // Check that the full trace computation scales with the number of nodes
+  // instead of paths in the DD due to the usage of a compute table
+  for (std::size_t numQubits = 1; numQubits <= 10; ++numQubits) {
+    auto dd = std::make_unique<dd::Package<>>(numQubits);
+    auto& computeTable = dd->getTraceComputeTable<dd::mNode>();
+    const auto hGate = dd->makeGateDD(dd::H_MAT, 0);
+    auto hKron = hGate;
+    for (std::size_t i = 0; i < numQubits - 1; ++i) {
+      hKron = dd->kronecker(hKron, hGate, 1);
+    }
+    dd->trace(hKron, numQubits);
+    auto stats = computeTable.getStats();
+    auto lookups = stats.lookups;
+    auto hits = stats.hits;
+    ASSERT_EQ(lookups, 2 * numQubits - 1);
+    ASSERT_EQ(hits, numQubits - 1);
```

Comment:
```suggestion
    const auto& stats = computeTable.getStats();
    ASSERT_EQ(stats.lookups, 2 * numQubits - 1);
    ASSERT_EQ(stats.hits, numQubits - 1);
```
Just a tiny bit more compact.